README.md
# Swipeable button

A swipeable button widget.

## Features

- Customizable size: you can manually specify the height of the button. It
  takes the width from its parent element.

- Make it either triggerable one time only or infinite times.

- The label shown on the button can be **any** widget

- It can match the style of your application: you can customize the color
  of the button and fully customize the thumb!

## Live Preview

You can check the example page [here](https://dthach4.github.io/swipeable_button/)!

## Getting started

This package has been built for **Flutter**.

Add `swipeable_button: ^0.2.0` to your `pubspec.yaml` dependencies.

You need to import the library in every file that uses this package by using
```dart
import 'package:swipeable_button/swipeable_button.dart';
```

## Usage

The `SwipeableButton` class provides named constructors so it's easier to start
using it.

### Basic usage

You can use the simplified constructor `SwipeableButton.simple` to create a
`SwipeableButton` widget.

```dart
SwipeableButton.simple(
  label: const Center(
    child: Text(
      "Delete",
      style: TextStyle(
        color: Colors.white,
        fontSize: 16.0,
      ),
    ),
  ),
  height: 40.0,
  minThumbWidth: 40.0,
  oneTime: true,
  color: Colors.red,
  thumbColor: Colors.orange,
  thumbIconColor: Colors.black,
  borderRadius: BorderRadius.circular(5.0),
  onSwipe: () {
    // Perform delete action.
  },
)
```

There is also a simplified constructor `SwipaebleButton.simpleSlidable` that
creates a `SwipeableButton` widget with a slidable thumb.

```dart
SwipeableButton.simpleSlidable(
  label: const Center(
    child: Text(
      "Delete",
      style: TextStyle(
        color: Colors.white,
        fontSize: 16.0,
      ),
    ),
  ),
  height: 40.0,
  minThumbWidth: 40.0,
  oneTime: true,
  color: Colors.red,
  thumbColor: Colors.orange,
  thumbIconColor: Colors.black,
  borderRadius: BorderRadius.circular(5.0),
  onSwipe: () {
    // Perform delete action.
  },
)
```

### Default usage

You can customize the `SwipeableButton` by using the default constructor.

```dart
SwipeableButton(
  height: 40.0,
  minThumbWidth: 40.0,
  oneTime: true,
  label: const Text("Delete"),
  color: Colors.red,
  borderRadius: BorderRadius.circular(5.0),
  thumbBuilder: (BuildContext context,
          double swipedFraction, bool isComplete) =>
      Padding(
        padding: const EdgeInsets.all(2.0),
        child: DecoratedBox(
          decoration: BoxDecoration(
              color: swipedFraction < 0.5
                ? Color.lerp(Colors.red, Colors.yellow, 2.0 * swipedFraction)
                : Color.lerp(Colors.yellow, Colors.green, 2.0 * swipedFraction - 1.0),
              borderRadius: BorderRadius.circular(16.0)),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.end,
            children: <Widget>[
              SizedBox(
                width: 36.0,
                child: isComplete
                    ? const Icon(Icons.sentiment_very_satisfied,
                        color: Colors.white)
                    : const Icon(Icons.sentiment_very_dissatisfied,
                        color: Colors.white),
              ),
            ],
          ),
        ),
      ),
  onSwipe: () {
    // Perform delete action.
  },
)
```

CHANGELOG.md
## 0.2.0

 * Added class `SwipeableButtonSimpleSlidableThumb` which is used for the
   `SwipeableButton.simpleSlidable` named constructor.
 * Docs: Fixed documentation and example.

## 0.1.1

* Added class `SwipeableButtonSimpleThumb` which is used for the
  `SwipeableButton.simple` named constructor.
* Docs: Fixed documentation and example.

## 0.1.0

* The previous `SwipeableButton` can now be built via the named constructor
  `SwipeableButton.simple`

* The default `SwipeableButton` constructor now takes a `thumbBuilder` function
  instead of a `thumbColor` parameter, so the thumb widget can be fully
  customized

## 0.0.1

Initial version

* The button can be triggered either one time only or infinite times.

* The label shown on the button can be any widget.

* The button's border radius, the button color and the thumb's color can be
  customized.

* Height and minimum thumb width can be specified.

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:swipeable_button/swipeable_button.dart';

void main() => runApp(const SwipeableButtonExample1());

@immutable
class SwipeableButtonExample1 extends StatelessWidget {
  const SwipeableButtonExample1({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => const MaterialApp(
        home: SwipeableButtonExample1HomePage(),
      );
}

@immutable
class SwipeableButtonExample1HomePage extends StatelessWidget {
  const SwipeableButtonExample1HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: const Text("Swipeable Button Example"),
        ),
        body: Center(
          child: SizedBox(
            width: 300.0,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton.simple(
                      label: const Center(
                          child: Text("Simple blue button",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      color: Colors.blue.shade400,
                      thumbColor: Colors.blue.shade800,
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton.simple(
                      label: const Center(
                          child: Text("Simple red button",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      color: Colors.red.shade400,
                      thumbColor: Colors.red.shade800,
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton(
                      label: const Center(
                          child: Text("Red to green thumb",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      color: Colors.blue.shade400,
                      thumbBuilder: (BuildContext context,
                              double swipedFraction, bool isComplete) =>
                          SwipeableButtonSimpleThumb(
                            color: Color.lerp(
                                    Colors.red, Colors.green, swipedFraction) ??
                                Colors.red,
                            iconColor: Colors.black,
                            minWidth: 40.0,
                            isComplete: isComplete,
                          ),
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton.simpleSlidable(
                      label: const Center(
                          child: Text("Slidable blue button",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      color: Colors.blue.shade400,
                      thumbColor: Colors.blue.shade800,
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton.simpleSlidable(
                      label: const Center(
                          child: Text("Slidable red button",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      height: 40.0,
                      minThumbWidth: 40.0,
                      oneTime: true,
                      color: Colors.red.shade400,
                      thumbColor: Colors.red.shade800,
                      borderRadius: BorderRadius.circular(5.0),
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: SwipeableButton(
                      label: const Center(
                          child: Text("Custom thumb",
                              style: TextStyle(
                                  color: Colors.white, fontSize: 16.0))),
                      color: Colors.blue.shade400,
                      minThumbWidth: 40.0,
                      thumbBuilder: (BuildContext context,
                              double swipedFraction, bool isComplete) =>
                          Padding(
                            padding: const EdgeInsets.all(2.0),
                            child: DecoratedBox(
                              decoration: BoxDecoration(
                                  color: swipedFraction < 0.5
                                      ? Color.lerp(Colors.red, Colors.yellow,
                                          2.0 * swipedFraction)
                                      : Color.lerp(Colors.yellow, Colors.green,
                                          2.0 * swipedFraction - 1.0),
                                  borderRadius: BorderRadius.circular(16.0)),
                              child: Row(
                                mainAxisAlignment: MainAxisAlignment.end,
                                children: <Widget>[
                                  SizedBox(
                                    width: 36.0,
                                    child: isComplete
                                        ? const Icon(
                                            Icons.sentiment_very_satisfied,
                                            color: Colors.white)
                                        : const Icon(
                                            Icons.sentiment_very_dissatisfied,
                                            color: Colors.white),
                                  ),
                                ],
                              ),
                            ),
                          ),
                      onSwipe: () {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text("Thank you for swiping!"),
                        ));
                      }),
                ),
              ],
            ),
          ),
        ),
      );
}
